### PR TITLE
IL-2863 open swap v2 draft

### PR DESCRIPTION
**Pull request by mario-ipor**

### CODE REVIEW

File: contracts/router/IporProtocolRouter.sol [hunk 1]
```diff
@@ -28,6 +32,27 @@ contract IporProtocolRouter is UUPSUpgradeable, AccessControl {
             sig == IAmmSwapsLens.getSwaps.selector
         ) {
             return AMM_SWAPS_LENS;
+        } else if (
```
Comment (amydlarz) [inline contracts/router/IporProtocolRouter.sol:35]: what about reentrance?

Comment (amydlarz) [inline contracts/router/IporProtocolRouter.sol:35]: what about pause?

File: contracts/amm/AmmOpenSwapService.sol [hunk 2]
```diff
@@ -0,0 +1,811 @@
...
+
+        MiltonTypes.IporSwapIndicator memory indicator = MiltonTypes.IporSwapIndicator(
+            bosStruct.accruedIpor.indexValue,
+            bosStruct.accruedIpor.ibtPrice,
+            IporMath.division(bosStruct.notional * Constants.D18, bosStruct.accruedIpor.ibtPrice),
+            quoteValue
+        );
+
+        AmmTypes.NewSwap memory newSwap = AmmTypes.NewSwap(
+            msg.sender,
```
Comment (pete-ipor) [inline contracts/amm/AmmOpenSwapService.sol:515]: i think we want in this place `context.behalfOf`?

File: contracts/amm/AmmOpenSwapService.sol [hunk 3]
```diff
@@ -0,0 +1,811 @@
...
```
Comment (pete-ipor) [inline contracts/amm/AmmOpenSwapService.sol:617]: `context.behalfOf`??

File: contracts/amm/AmmOpenSwapService.sol [hunk 4]
```diff
@@ -0,0 +1,811 @@
...
+
+        uint256 wadTotalAmount = IporMath.convertToWad(totalAmount, IERC20MetadataUpgradeable(asset).decimals());
+
+        require(
+            wadTotalAmount > _liquidationDepositAmountWad + _iporPublicationFee,
+            MiltonErrors.TOTAL_AMOUNT_LOWER_THAN_FEE
+        );
+
+        (uint256 collateral, uint256 notional, uint256 openingFeeAmount) = IporSwapLogic.calculateSwapAmount(
+            28,
```
Comment (pete-ipor) [inline contracts/amm/AmmOpenSwapService.sol:666]: what with 60/90?

File: contracts/amm/AmmOpenSwapService.sol [hunk 5]
```diff
@@ -0,0 +1,811 @@
...
+        if (maxCollateralPerLeg > 0) {
+            riskIndicators.maxLeveragePerLeg = leverageInRange(
+                IporMath.division(maxNotionalPerLeg * Constants.D18, maxCollateralPerLeg)
+            );
+        } else {
+            riskIndicators.maxLeveragePerLeg = _minLeverage;
+        }
+    }
+
+    function leverageInRange(uint256 leverage) internal view returns (uint256) {
```
Comment (pete-ipor) [inline contracts/amm/AmmOpenSwapService.sol:732]: `_leverageInRange`

File: contracts/amm/MiltonStorage.sol [hunk 6]
```diff
@@ -335,10 +305,12 @@ contract MiltonStorage is
...
     }
 
-    function updateStorageWhenOpenSwapPayFixed(
-        AmmTypes.NewSwap memory newSwap,
-        uint256 cfgIporPublicationFee
-    ) external override onlyMilton returns (uint256) {
+    function updateStorageWhenOpenSwapPayFixed(AmmTypes.NewSwap memory newSwap, uint256 cfgIporPublicationFee)
+        external
+        override
+        onlyMilton
```
Comment (pete-ipor) [inline contracts/amm/MiltonStorage.sol:311]: `onlyRouter`

File: contracts/amm/MiltonStorage.sol [hunk 7]
```diff
@@ -356,10 +328,12 @@ contract MiltonStorage is
...
     }
 
-    function updateStorageWhenOpenSwapReceiveFixed(
-        AmmTypes.NewSwap memory newSwap,
-        uint256 cfgIporPublicationFee
-    ) external override onlyMilton returns (uint256) {
+    function updateStorageWhenOpenSwapReceiveFixed(AmmTypes.NewSwap memory newSwap, uint256 cfgIporPublicationFee)
+        external
+        override
+        onlyMilton
```
Comment (pete-ipor) [inline contracts/amm/MiltonStorage.sol:334]: `onlyRouter`

File: contracts/amm/AmmOpenSwapService.sol [hunk 8]
```diff
@@ -0,0 +1,819 @@
...
+import "./libraries/types/AmmMiltonTypes.sol";
+import "../libraries/errors/MiltonErrors.sol";
+import "./libraries/IporSwapLogic.sol";
+
+contract AmmOpenSwapService is IAmmOpenSwapService {
+    using Address for address;
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    /// @dev represented in 18 decimals
+    uint256 internal immutable _iporPublicationFee;
```
Comment (kris-ipor) [inline contracts/amm/AmmOpenSwapService.sol:26]: maybe we should have configurations per asset, what do you think?

File: contracts/amm/AmmOpenSwapService.sol [hunk 9]
```diff
@@ -0,0 +1,819 @@
...
+        uint256 acceptableFixedInterestRate,
+        uint256 leverage
+    ) external override returns (uint256) {
+        Context memory context = Context({
+            onBehalfOf: onBehalfOf,
+            asset: _usdt,
+            ammStorage: _usdtAmmStorage,
+            ammTreasury: _usdtAmmTreasury,
+            maturity: 28,
+            spreadMethodSig: "calculatePayFixed28Days((address,uint256,uint256,uint256,int256,uint256,uint256,uint256,uint256,uint256))"
```
Comment (kris-ipor) [inline contracts/amm/AmmOpenSwapService.sol:158]: will it be replaced by signatures from interface later?

File: contracts/amm/AmmOpenSwapService.sol [hunk 10]
```diff
@@ -0,0 +1,819 @@
...
+        uint256 totalAmount,
+        uint256 acceptableFixedInterestRate,
+        uint256 leverage
+    ) external override returns (uint256) {
+        Context memory context = Context({
+            onBehalfOf: onBehalfOf,
+            asset: _usdc,
+            ammStorage: _usdcAmmStorage,
+            ammTreasury: _usdcAmmTreasury,
+            maturity: 28,
```
Comment (kris-ipor) [inline contracts/amm/AmmOpenSwapService.sol:261]: as we agreed risk management oracle doesn't match swap duration by number of days,
` @param duration swap duration, 0 = 28 days, 1 = 60 days, 2 = 90 days`

File: contracts/amm/libraries/OpenSwapLogic.sol [hunk 11]
```diff
@@ -0,0 +1,6 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity 0.8.16;
+
+library OpenSwapLogic {
```
Comment (kris-ipor) [inline contracts/amm/libraries/OpenSwapLogic.sol:4]: empty library

**Review by kris-ipor — CHANGES_REQUESTED** [review]

File: contracts/amm/AmmOpenSwapService.sol (hunk 9 above)
Comment (mario-ipor) [reply contracts/amm/AmmOpenSwapService.sol:158]: Yes, when interface came with Spread implementation

File: contracts/amm/AmmOpenSwapService.sol (hunk 8 above)
Comment (mario-ipor) [reply contracts/amm/AmmOpenSwapService.sol:26]: good idea, I will change

**Review by kris-ipor — APPROVED** [review]

**Pull request merged**